test(electrum): Test sync in reorg and no-reorg situations

LagginTimes opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->
Closes #1125.

### Description

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->
Add new test for `bdk_electrum` to make sure previously unconfirmed transactions get confirmed again in both reorg and no-reorg situations.

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->
* Added `wait_until_electrum_sees_txid` method to `TestEnv`.
* `wait_until_electrum_sees_block` now has a `Duration` input for timeout.
* Removed exponential polling delay in `wait_until_electrum_sees_block`.
* Added `test_sync` to `bdk_electrum` to make sure previously unconfirmed transactions get confirmed in both reorg and no-org situations.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -22,10 +23,34 @@ fn get_balance(
     Ok(balance)
 }
 
+fn electrum_sync<I>(
```
**Nit:** according to [standard naming conventions in Rust](https://rust-lang.github.io/api-guidelines/naming.html#names-use-a-consistent-word-order-c-word-order), let's start with the verb for this function (since this is how we name things in BDK, let me know if you find an exception though). I.e. `sync_with_electrum`.

evanlinjin replied:
@LagginTimes I think you attempted to change this, but you ended up changing the wrong thing 😅

LagginTimes replied:
Oops.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -22,10 +23,34 @@ fn get_balance(
 pub fn test_update_tx_graph_without_keychain() -> anyhow::Result<()> {
     let env = TestEnv::new()?;
-    let electrum_client = electrum_client::Client::new(env.electrsd.electrum_url.as_str())?;
+    let electrum_client = Client::new(env.electrsd.electrum_url.as_str())?;
```
**Nit:** Ideally, we don't change lines of code that we don't need to touch. Same as line `151`. In this file, could we use the full module path to refer to `Client`?

However, if you see a problem with using the full path, I think it makes sense to include this change in a separate commit.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -238,18 +263,13 @@ pub fn test_update_tx_graph_stop_gap() -> anyhow::Result<()> {
-/// 2. Send a tx.
-/// 3. Mine extra block to confirm sent tx.
-/// 4. Check [`Balance`] to ensure tx is confirmed.
+/// Ensure that [`BdkElectrumClient`] can sync properly in both reorg and no-reorg situations.
```
This doc comment is a little ambiguous. We may need to expand on "sync properly" here and describe the scenario in more detail. I think you did a better job describing the test in the PR description than in this doc comment.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -22,10 +23,34 @@ fn get_balance(
 
+fn electrum_sync<I>(
+    client: &BdkElectrumClient<Client>,
+    spks: &[ScriptBuf],
```
**Nit:** there is no point having a reference here, since we are cloning this later on (with `.to_owned()`).

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -22,10 +23,34 @@ fn get_balance(
+) -> anyhow::Result<SyncResult>
+where
+    I: Indexer,
+    I::ChangeSet: Default + Merge,
```
**Nit:** This removes the unnecessary reference and allows more flexibility for what we can input in `spks`.

```suggestion
fn electrum_sync<I, Spks>(
    client: &BdkElectrumClient<Client>,
    spks: Spks,
    batch_size: usize,
    chain: &mut LocalChain,
    graph: &mut IndexedTxGraph<ConfirmationBlockTime, I>,
) -> anyhow::Result<SyncResult>
where
    I: Indexer,
    I::ChangeSet: Default + Merge,
    Spks: IntoIterator<Item = ScriptBuf>,
    Spks::IntoIter: ExactSizeIterator + Send + 'static,

```

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -22,10 +23,34 @@ fn get_balance(
+fn electrum_sync<I>(
+    client: &BdkElectrumClient<Client>,
+    spks: &[ScriptBuf],
+    batch_size: usize,
```
I think it's okay to hard-code the batch size and have it as a constant at the top of the file (instead of having it as an input to this function). The value is not relevant for what we are testing and not having it slightly increases readability for the tests.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -272,32 +292,91 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
-    // Create transaction that is tracked by our receiver.
+    // Broadcast transaction to mempool.
     env.send(&addr_to_track, SEND_AMOUNT)?;
+    env.wait_until_electrum_sees_block()?;
```
We are sending tx to mempool, but we are waiting to see a block? Is there a method that waits for electrum to see a certain tx in mempool? If not, we should add it.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -272,32 +292,91 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
+            ..Balance::default()
+        },
+        "initial balance must be correct",
+    );
```
Since we have a tx in mempool, shouldn't we have an unconfirmed balance?

If the intention is to sync BEFORE seeing the tx in mempool, it should be made more obvious in code.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -272,32 +292,91 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
+            confirmed: Amount::from_sat(0),
+            ..Balance::default()
+        },
+    );
```
This result is assuming that the receiving structures have not seen the tx in mempool yet.

If the receiving structures are aware of the tx being in mempool, there should be an unconfirmed balance here.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -272,32 +292,91 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
+        "initial balance must be correct",
     );
 
     for tx in recv_graph.graph().full_txs() {
```
We need a comment before the for loop to explain what is being tested in the for loop.

evanlinjin reviewed (changes requested):
Good work with this. Moving forward, I recommend improving the docs and adding more clarity to the testing logic. I also included a bunch of nits for you to consider including.

evanlinjin commented:
The commit also adds a method to the `TestEnv` (which is not documented). Please either have this change as a separate commit, or mention the change in the commit.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -186,6 +186,22 @@ impl TestEnv {
         }
     }
 
+    /// This method waits for Electrum to see a transaction with given `txid`.
```
```suggestion
    /// This method waits for Electrsd to see a transaction with given `txid`.
```

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -186,6 +186,22 @@ impl TestEnv {
+        loop {
+            if self.electrsd.client.transaction_get(&txid).is_ok() {
+                return Ok(());
+            }
```
This will create a situation where we may loop forever.

Are we able to figure out if the error is a not-exist error? Ideally, the `rust_esplora` crate has something to figure this out easily. If not, we should upstream some changes later.

Also, would be good to add a timeout as an input. Just in case the test fails (the txid never shows up). We at least want to exit the test!

evanlinjin replied:
The timeout thing can be added in a separate PR. Can also add timeouts to the other wait methods.

LagginTimes replied:
The error returned is: `Err(Protocol(String("missing transaction")))`

LagginTimes committed:
feat(testenv): Added `bdk_electrum` wait for Txid method

Added `wait_until_electrum_sees_txid` method to `TestEnv`. Both
`bdk_electrum` wait methods now have a `timeout` option. Removed
the exponential polling delay in lieu of a fixed delay inside the
`bdk_electrum` wait methods.

LagginTimes committed:
test(electrum): Test sync in reorg and no-reorg situations

Add test for `bdk_electrum` to make sure previously unconfirmed
transactions get confirmed again in both reorg and no-reorg
situations.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -168,22 +168,36 @@ impl TestEnv {
     /// This method waits for the Electrum notification indicating that a new block has been mined.
-    pub fn wait_until_electrum_sees_block(&self) -> anyhow::Result<()> {
+    /// `timeout` is the maximum amount of seconds we want to wait for a response from Electrsd.
+    pub fn wait_until_electrum_sees_block(&self, timeout: u32) -> anyhow::Result<()> {
```
Can we please make `timeout` a `Duration`?

You can instantiate an `Instant` before the loop and check `Instant::elapsed` against `timeout` within the loop to determine if we need to break.

https://doc.rust-lang.org/std/time/struct.Instant.html#method.elapsed

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -168,22 +168,42 @@ impl TestEnv {
+    }
 
-            if delay.as_millis() < 512 {
-                delay = delay.mul_f32(2.0);
```
Is there a reason why we got rid of the exponential backoff strategy?

evanlinjin replied:
As mentioned in DMs, I'm okay with removing this, but can we please document this removal in the commit description? Thanks!

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -271,35 +303,88 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
+        &mut recv_graph,
+    )?;
+
+    // Check if balance is zero when transaction exists only in mempool.
```
Might need to update this comment now, since we have a pending balance.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -271,35 +303,88 @@ fn scan_detects_confirmed_tx() -> anyhow::Result<()> {
+        "balance must be correct",
     );
 
+    // Check to see if we have the floating txouts available from our transactions' previous outputs
+    // in order to calculate transaction fees.
```
Since we are testing this, might make sense to mention this in the test's doc comments.

evanlinjin reviewed (changes requested):
LGTM. Once these are addressed and changelog is added to PR description, I am happy to merge.

evanlinjin reviewed (approved):
ACK 2c0bc45ecf0241abc8bf9d72c8bbeded53d32a45

ValuedMammal commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -407,24 +490,21 @@ fn tx_can_become_unconfirmed_after_reorg() -> anyhow::Result<()> {
         assert_eq!(
             get_balance(&recv_chain, &recv_graph)?,
             Balance {
+                trusted_pending: SEND_AMOUNT * depth as u64,
```
I agree this line should be here. It made me wonder why after #1416 the assertion failed, but I think the difference here is that we're updating last-seen during `sync_with_electrum` as we should.

see related #1550
